src/main/java/gift/dao/ProductDao.java

```diff
@@ -0,0 +1,78 @@
+package gift.dao;
+
+import gift.domain.Product;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public class ProductDao implements ProductRepository {
+
+    private final JdbcTemplate jdbcTemplate;
+
+    public ProductDao(DataSource dataSource) {
+        this.jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    @Override
+    public Product save(Product product) {
+        String sql = "INSERT INTO product (name, price, image_url) VALUES(?, ?, ?)";
+        jdbcTemplate.update(sql, product.getName(), product.getPrice(), product.getImageUrl());
+        return product;
+    }
+
+    @Override
+    public Optional<Product> findById(Long id) {
+        String sql = "SELECT * FROM product WHERE id = ?";
+        return Optional.ofNullable(
+                jdbcTemplate.queryForObject(
+                        sql,
+                        (rs, rowNum) -> new Product(
+                                rs.getLong("id"),
+                                rs.getString("name"),
+                                rs.getInt("price"),
+                                rs.getString("image_url")
+                        ),
+                        id
+                )
+        );
```

Comment:
`.queryForObject()` 사용시에 데이터가 없으면 EmptyResultDataAccessException 에러가 나지 않았을까요?!  데이터가 없는 경우에 Optional 을 반환하기 위해서는 아래 코드로 변경이 필요할 것 같습니다. 
```suggestion
                return jdbcTemplate.query(
                sql,
                (rs, rowNum) -> new Product(
                    rs.getLong("id"),
                    rs.getString("name"),
                    rs.getInt("price"),
                    rs.getString("image_url")
                ),
                id
            )
            .stream()
            .findFirst();
```

fyi; JdbcClient queryForObject 내부에서 수행되는 코드
```java
    public static <T> T requiredSingleResult(@Nullable Collection<T> results) throws IncorrectResultSizeDataAccessException {
        if (CollectionUtils.isEmpty(results)) { // 🔥🔥🔥 데이터가 null 인경우, 에러 발생
            throw new EmptyResultDataAccessException(1);
        } else if (results.size() > 1) {
            throw new IncorrectResultSizeDataAccessException(1, results.size());
        } else {
            return results.iterator().next();
        }
    }
```